albatross/tune.h

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright (C) 2018 Swift Navigation Inc.
+ * Contact: Swift Navigation <dev@swiftnav.com>
+ *
+ * This source is subject to the license found in the file 'LICENSE' which must
+ * be distributed together with this source. All other rights reserved.
+ *
+ * THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND,
+ * EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
+ */
+
+#ifndef ALBATROSS_TUNE_H
+#define ALBATROSS_TUNE_H
+
+#include <map>
+#include <vector>
+#include "core/model.h"
+#include "nlopt.hpp"
+
+namespace albatross {
+
+std::vector<ParameterValue> transform_parameters(
+    const std::vector<ParameterValue>& x) {
+  std::vector<ParameterValue> transformed(x.size());
+  for (std::size_t i = 0; i < x.size(); i++) {
+    transformed[i] = log(x[i]);
+  }
+  return transformed;
+}
+
+std::vector<ParameterValue> inverse_parameters(
+    const std::vector<ParameterValue>& x) {
+  std::vector<ParameterValue> inverted(x.size());
+  for (std::size_t i = 0; i < x.size(); i++) {
+    inverted[i] = exp(x[i]);
+  }
+  return inverted;
+}
+template <class Predictor>
+using TuningMetric = std::function<double(const RegressionDataset<Predictor> &,
+                                          RegressionModel<Predictor> *)>;
+
+template <class Predictor>
+struct TuneModelConfg {
+  RegressionModelCreator<Predictor> model_creator;
+  RegressionDataset<Predictor> dataset;
+  TuningMetric<Predictor> metric;
+
+  TuneModelConfg(const RegressionModelCreator<Predictor> &model_creator_,
+                 const RegressionDataset<Predictor> &dataset_,
+                 const TuningMetric<Predictor> &metric_)
+      : model_creator(model_creator_), dataset(dataset_), metric(metric_){};
+};
+
+template <class Predictor>
+double objective_function(const std::vector<double> &x,
+                          std::vector<double> &grad, void *void_tune_config) {
```

Comment:
Is `grad` const? or is this the nlopt api?

Reply:
That's the `nlopt` api.  If the optimization algorithm you choose requires gradients you're expected to have filled them in hence the non const.